File: testapp/wincap.swift
// Captures one window with PrintWindow(PW_RENDERFULLCONTENT) and writes a BMP.
//
//   wincap.exe "<title substring>" <out.bmp>
//
// Exit codes, which are the interface: 0 captured and the image has content,
// 1 no visible window matched, 2 bad arguments, 3 captured but the image is
// entirely black.
//
// Why it exists. screenshot.zsh captures windows with ffmpeg's gdigrab, which
// is GDI/BitBlt, and BitBlt needs a redirection surface to copy from. Two kinds
// of window here do not have one:
//
//   - GTK on Windows with Direct Composition enabled (`-GPU 2`). GDK creates
//     the toplevel with WS_EX_NOREDIRECTIONBITMAP; measured 2026-08-29 as
//     exstyle 0x00200000, against 0x00000100 for a window that captures fine.
//     The style cannot be removed afterwards -- SetWindowLongPtrW returns 0
//     with GetLastError 87, ERROR_INVALID_PARAMETER.
//   - A WSLg window seen from the Windows side, which is WS_EX_LAYERED
//     (0x00080100). gdigrab given the exact title finds it and copies nothing:
//     45 non-black pixels out of 155,775 sampled.
//
// This is not a property of Direct Composition or of D3D. P6 presents a D3D11
// composition swapchain and gdigrab captures its window perfectly, because its
// HWND keeps its redirection surface. What matters is the window, not what
// draws into it.
//
// PW_RENDERFULLCONTENT asks DWM to render the window instead of copying a
// surface, which is the documented route for exactly this case. Measured: 93.0%
// non-black for the DComp GTK window, 92.6% for the WSLg one, both showing the
// complete window.
//
// The black-pixel count is the point, not a nicety. PrintWindow returns TRUE
// while producing an entirely black bitmap -- the same failure gdigrab has
// above -- so a caller reading only the exit status would be told an empty
// image was a success. Exit 3 exists for that.
//
// 以 PrintWindow(PW_RENDERFULLCONTENT) 擷取單一視窗並寫出 BMP。
//
// 結束碼即為它的介面：0 成功且影像有內容、1 找不到符合的可見視窗、2 參數錯誤、
// 3 擷取到了但影像整片全黑。
//
// 存在的理由。screenshot.zsh 以 ffmpeg 的 gdigrab 擷取視窗，那是 GDI/BitBlt，而 BitBlt 需要
// 有 redirection surface 可供複製。此處有兩種視窗沒有：
//
//   - 啟用 Direct Composition 的 Windows GTK（`-GPU 2`）。GDK 以
//     WS_EX_NOREDIRECTIONBITMAP 建立頂層視窗；2026-08-29 實測 exstyle 為 0x00200000，
//     而可正常擷取的視窗為 0x00000100。該樣式建立後無法移除——SetWindowLongPtrW 回傳 0，
//     GetLastError 為 87（ERROR_INVALID_PARAMETER）。
//   - 從 Windows 這側看到的 WSLg 視窗，其樣式為 WS_EX_LAYERED（0x00080100）。即使給
//     gdigrab 完全正確的標題，它找得到卻什麼也複製不到：取樣的 155,775 個像素中僅 45 個非黑。
//
// 這並非 Direct Composition 或 D3D 的性質。P6 呈現的是 D3D11 的 composition swapchain，
// gdigrab 卻能完美擷取其視窗，因為它的 HWND 保有 redirection surface。決定成敗的是「視窗」，
// 不是「誰在往裡面畫」。
//
// PW_RENDERFULLCONTENT 要求 DWM 重新繪製該視窗，而非複製既有表面，正是為這種情況所設計的
// 途徑。實測：DComp 的 GTK 視窗非黑像素達 93.0%，WSLg 視窗為 92.6%，兩者都完整呈現視窗。
//
// 黑色像素計數才是重點，不是附加品。PrintWindow 會在產出全黑點陣圖的同時回傳 TRUE——與上文
// gdigrab 的失敗形態相同——因此只讀結束碼的呼叫端會把空白影像當成成功。結束碼 3 就是為此存在。

import Foundation
import WinSDK

let args = CommandLine.arguments
guard args.count >= 3 else {
    print("usage: wincap.exe <title substring> <out.bmp>")
    exit(2)
}
let needle = args[1].lowercased()
let outPath = args[2]

// Matching is a case-insensitive SUBSTRING, deliberately, because the exact
// title is not always what the caller knows. WSLg appends " (Ubuntu)" and, when
// its GPU path is degraded, prepends "[WARN:COPY MODE] " -- so the window the
// caller asked for by app name is neither of those strings. gdigrab matches
// exactly and fails on precisely that.
// 比對刻意採用「不分大小寫的子字串」，因為呼叫端未必知道確切標題。WSLg 會附加
// 「 (Ubuntu)」，而當其 GPU 路徑降級時還會前置「[WARN:COPY MODE] 」——於是呼叫端以 app
// 名稱指定的那個視窗，兩種字串都不是。gdigrab 採精確比對，正是敗在這一點。
var target: HWND?
func title(of hwnd: HWND) -> String {
    var buffer = [WCHAR](repeating: 0, count: 512)
    guard GetWindowTextW(hwnd, &buffer, 512) > 0 else { return "" }
    return String(decodingCString: buffer, as: UTF16.self)
}
let finder: @convention(c) (HWND?, LPARAM) -> WindowsBool = { hwnd, _ in
    guard let hwnd, IsWindowVisible(hwnd), target == nil else { return true }
    let name = title(of: hwnd)
    if !name.isEmpty, name.lowercased().contains(needle) {
        target = hwnd
        print("window: \(name)")
        return false
    }
    return true
}

// A WSLg HWND is published asynchronously after the Linux process starts.
// One enumeration races that publication: the hardware renderer takes just
// long enough that a capture requested after one second can see no window,
// while the same capture a few seconds later succeeds. Poll Windows rather
// than falling back to a desktop image; success must always mean the named
// window itself was captured.
//
// WSLg 的 HWND 會在 Linux process 啟動後非同步發布。只列舉一次會與發布競速：硬體
// renderer 啟動稍慢，使一秒後的擷取可能還看不到視窗，數秒後同一擷取卻成功。此處輪詢
// Windows，而不退回 desktop image；成功必須始終代表擷取到指定視窗本身。
let deadline = GetTickCount64() + 5_000
repeat {
    target = nil
    _ = EnumWindows(finder, 0)
    if target != nil { break }
    Sleep(250)
} while GetTickCount64() < deadline

guard let hwnd = target else {
    print("no visible window matched \"\(needle)\"")
    exit(1)
}

var rect = RECT()
GetWindowRect(hwnd, &rect)
let width = Int(rect.right - rect.left)
let height = Int(rect.bottom - rect.top)
guard width > 0, height > 0 else {
    print("window has no area")
    exit(1)
}

// Reported so a caller can see WHY a capture failed rather than only that it
// did. 0x00200000 is WS_EX_NOREDIRECTIONBITMAP and 0x00080000 is WS_EX_LAYERED;
// either explains a black gdigrab capture, and neither is visible from ffmpeg.
// 一併回報，讓呼叫端看得到擷取「為何」失敗，而不只是「失敗了」。0x00200000 是
// WS_EX_NOREDIRECTIONBITMAP，0x00080000 是 WS_EX_LAYERED；兩者都能解釋 gdigrab 拍出全黑，
// 而從 ffmpeg 一側都看不見。
let ex = GetWindowLongPtrW(hwnd, GWL_EXSTYLE)
let exHex = String(UInt64(bitPattern: Int64(ex)), radix: 16)
print("size: \(width)x\(height)  exstyle: 0x\(exHex)")

let screenDC = GetDC(nil)
let memDC = CreateCompatibleDC(screenDC)
var info = BITMAPINFO()
info.bmiHeader.biSize = DWORD(MemoryLayout<BITMAPINFOHEADER>.size)
info.bmiHeader.biWidth = LONG(width)
info.bmiHeader.biHeight = LONG(-height)     // top-down
info.bmiHeader.biPlanes = 1
info.bmiHeader.biBitCount = 32
info.bmiHeader.biCompression = DWORD(BI_RGB)

var bits: UnsafeMutableRawPointer?
guard let dib = CreateDIBSection(memDC, &info, UINT(DIB_RGB_COLORS), &bits, nil, 0),
      let pixels = bits else {
    print("CreateDIBSection failed")
    exit(1)
}
let old = SelectObject(memDC, dib)

// PW_RENDERFULLCONTENT is 0x00000002. Not every SDK mapping exposes it as a
// constant, so it is written out rather than assumed present.
// PW_RENDERFULLCONTENT 為 0x00000002。並非每一種 SDK 映射都會將它提供為常數，因此直接寫出
// 數值，而不假設它存在。
let ok = PrintWindow(hwnd, memDC, UINT(0x0000_0002))

let buffer = pixels.bindMemory(to: UInt8.self, capacity: width * height * 4)
var nonBlack = 0
for index in stride(from: 0, to: width * height * 4, by: 4) {
    if buffer[index] > 8 || buffer[index + 1] > 8 || buffer[index + 2] > 8 { nonBlack += 1 }
}
let tenths = nonBlack * 1000 / (width * height)
print("PrintWindow: \(ok)  non-black: \(nonBlack)/\(width * height) (\(tenths / 10).\(tenths % 10)%)")

// BMP, because it needs no encoder: a 14-byte file header, the 40-byte info
// header already built above, then the rows bottom-up. The caller converts to
// PNG with the ffmpeg it already depends on.
// 選用 BMP，因為它不需要任何編碼器：14 位元組的檔頭、上方已建好的 40 位元組資訊標頭，
// 接著是由下而上的各列像素。呼叫端再以它本來就依賴的 ffmpeg 轉為 PNG。
let rowBytes = width * 4
let pixelBytes = rowBytes * height
var file = [UInt8]()
file.reserveCapacity(54 + pixelBytes)
func append32(_ value: UInt32) { for shift in [0, 8, 16, 24] { file.append(UInt8((value >> UInt32(shift)) & 0xFF)) } }
func append16(_ value: UInt16) { file.append(UInt8(value & 0xFF)); file.append(UInt8(value >> 8)) }
file.append(0x42); file.append(0x4D)                  // "BM"
append32(UInt32(54 + pixelBytes)); append32(0); append32(54)
append32(40); append32(UInt32(width)); append32(UInt32(height))
append16(1); append16(32); append32(0); append32(UInt32(pixelBytes))
append32(2835); append32(2835); append32(0); append32(0)
for row in stride(from: height - 1, through: 0, by: -1) {
    let start = row * rowBytes
    file.append(contentsOf: UnsafeBufferPointer(start: buffer + start, count: rowBytes))
}

SelectObject(memDC, old)
DeleteObject(dib)
DeleteDC(memDC)
ReleaseDC(nil, screenDC)

do {
    try Data(file).write(to: URL(fileURLWithPath: outPath))
} catch {
    print("write failed: \(outPath)")
    exit(1)
}
print("wrote \(outPath)")
// Rejected on a FRACTION, not on a count.
//
// This read `nonBlack > 0` until 2026-09-01, and on that day it passed a capture
// of 1359 pixels out of 648528 -- 0.2%, effectively black. screenshot.zsh then
// printed `captured from priority 1: wincap window "P40"`, which reads as a
// clean capture of the application. A real capture of the same window, measured
// the same day, is 93.0%. The two sit two orders of magnitude apart, so 2%
// separates them with room to spare while staying far below anything that has
// content in it.
//
// That the guard existed and still let a black image through is the point: it
// was testing whether ANY pixel survived, when the question is whether the
// window did. The threshold is named in the rejection line so a genuinely dark
// window can be argued with rather than guessed at.
//
// 以**比例**判定，而非以計數。
//
// 在 2026-09-01 之前這裡寫的是 `nonBlack > 0`，而就在那天，它讓一張 648528 中僅 1359 個非黑
// 像素的擷取通過——0.2%，實質上是全黑。screenshot.zsh 隨即印出
// `captured from priority 1: wincap window "P40"`，讀起來就是「乾淨地擷取到該應用程式」。同一天
// 對同一個視窗的真實擷取是 93.0%。兩者相差兩個數量級，因此 2% 能綽綽有餘地分開它們，同時仍遠低於
// 任何「有內容」的視窗。
//
// 重點在於：這道防護存在，卻仍讓全黑影像通過——它檢查的是「有沒有任何一個像素活下來」，而真正的
// 問題是「那個視窗有沒有活下來」。拒絕訊息中會寫出閾值，好讓「本來就很暗的視窗」可以被據理爭論，
// 而不是被猜測。
let minimumTenths = 20  // 2.0 per cent
if tenths < minimumTenths {
    print(
        "capture rejected: \(tenths / 10).\(tenths % 10)% non-black is below the "
            + "\(minimumTenths / 10)% minimum, so this is a black image rather than a window"
    )
    exit(3)
}
exit(0)

File: testapp/wincap.cpp
// Captures a named top-level window with Windows Graphics Capture and writes a
// 32-bit BMP. PrintWindow remains a fallback for hosts where WGC is unavailable.

#include <windows.h>
#include <dwmapi.h>
#include <d3d11.h>
#include <dxgi1_2.h>
#include <windows.graphics.capture.interop.h>
#include <windows.graphics.directx.direct3d11.interop.h>

#include <winrt/Windows.Foundation.h>
#include <winrt/Windows.Graphics.Capture.h>
#include <winrt/Windows.Graphics.DirectX.Direct3D11.h>

#include <algorithm>
#include <chrono>
#include <condition_variable>
#include <cstdint>
#include <cstring>
#include <cstdio>
#include <cwctype>
#include <fstream>
#include <mutex>
#include <optional>
#include <string>
#include <vector>

using namespace winrt;
namespace capture = winrt::Windows::Graphics::Capture;
namespace directx = winrt::Windows::Graphics::DirectX;
namespace direct3d = winrt::Windows::Graphics::DirectX::Direct3D11;

namespace {

struct Image {
    int width = 0;
    int height = 0;
    std::vector<std::uint8_t> bgra;
};

int nonBlackTenths(Image const& image);

std::wstring needle;
HWND target = nullptr;

std::wstring windowTitle(HWND window) {
    wchar_t value[512]{};
    int length = GetWindowTextW(window, value, static_cast<int>(std::size(value)));
    return length > 0 ? std::wstring(value, length) : std::wstring();
}

std::wstring lowercase(std::wstring value) {
    std::transform(value.begin(), value.end(), value.begin(), [](wchar_t character) {
        return static_cast<wchar_t>(std::towlower(character));
    });
    return value;
}

BOOL CALLBACK findWindow(HWND window, LPARAM) {
    if (target || !IsWindowVisible(window)) return TRUE;
    auto title = windowTitle(window);
    if (!title.empty() && lowercase(title).find(needle) != std::wstring::npos) {
        target = window;
        std::wprintf(L"window: %ls\n", title.c_str());
        return FALSE;
    }
    return TRUE;
}

com_ptr<ID3D11Device> makeD3DDevice() {
    UINT flags = D3D11_CREATE_DEVICE_BGRA_SUPPORT;
    D3D_FEATURE_LEVEL levels[] = {
        D3D_FEATURE_LEVEL_11_1,
        D3D_FEATURE_LEVEL_11_0,
        D3D_FEATURE_LEVEL_10_1,
        D3D_FEATURE_LEVEL_10_0,
    };
    com_ptr<ID3D11Device> device;
    com_ptr<ID3D11DeviceContext> context;
    D3D_FEATURE_LEVEL level{};
    check_hresult(D3D11CreateDevice(
        nullptr, D3D_DRIVER_TYPE_HARDWARE, nullptr, flags, levels,
        static_cast<UINT>(std::size(levels)), D3D11_SDK_VERSION,
        device.put(), &level, context.put()));
    return device;
}

direct3d::IDirect3DDevice makeWinRTDevice(ID3D11Device* device) {
    com_ptr<IDXGIDevice> dxgi;
    check_hresult(device->QueryInterface(dxgi.put()));
    com_ptr<IInspectable> inspectable;
    check_hresult(CreateDirect3D11DeviceFromDXGIDevice(dxgi.get(), inspectable.put()));
    return inspectable.as<direct3d::IDirect3DDevice>();
}

capture::GraphicsCaptureItem captureItemForWindow(HWND window) {
    auto itemInterop = get_activation_factory<capture::GraphicsCaptureItem, IGraphicsCaptureItemInterop>();
    capture::GraphicsCaptureItem item{nullptr};
    check_hresult(itemInterop->CreateForWindow(
        window, guid_of<capture::GraphicsCaptureItem>(), put_abi(item)));
    return item;
}

std::optional<Image> captureWithWGC(
    capture::GraphicsCaptureItem const& item,
    std::chrono::milliseconds timeout
) {

    auto d3dDevice = makeD3DDevice();
    auto context = [&] {
        com_ptr<ID3D11DeviceContext> value;
        d3dDevice->GetImmediateContext(value.put());
        return value;
    }();
    auto winrtDevice = makeWinRTDevice(d3dDevice.get());
    auto size = item.Size();
    if (size.Width <= 0 || size.Height <= 0) return std::nullopt;

    auto pool = capture::Direct3D11CaptureFramePool::CreateFreeThreaded(
        winrtDevice, directx::DirectXPixelFormat::B8G8R8A8UIntNormalized, 2, size);
    auto session = pool.CreateCaptureSession(item);
    std::mutex mutex;
    std::condition_variable condition;
    capture::Direct3D11CaptureFrame captured{nullptr};
    auto token = pool.FrameArrived([&](auto const& sender, auto const&) {
        std::lock_guard lock(mutex);
        captured = sender.TryGetNextFrame();
        condition.notify_one();
    });

    session.StartCapture();
    std::optional<Image> result;
    auto deadline = std::chrono::steady_clock::now() + timeout;
    while (std::chrono::steady_clock::now() < deadline) {
        capture::Direct3D11CaptureFrame frame{nullptr};
        {
            std::unique_lock lock(mutex);
            condition.wait_until(lock, deadline, [&] { return captured != nullptr; });
            if (!captured) break;
            frame = captured;
            captured = nullptr;
        }

        auto contentSize = frame.ContentSize();
        auto access = frame.Surface().as<::Windows::Graphics::DirectX::Direct3D11::IDirect3DDxgiInterfaceAccess>();
        com_ptr<ID3D11Texture2D> source;
        check_hresult(access->GetInterface(__uuidof(ID3D11Texture2D), source.put_void()));
        D3D11_TEXTURE2D_DESC description{};
        source->GetDesc(&description);
        description.Width = static_cast<UINT>(contentSize.Width);
        description.Height = static_cast<UINT>(contentSize.Height);
        description.Usage = D3D11_USAGE_STAGING;
        description.BindFlags = 0;
        description.CPUAccessFlags = D3D11_CPU_ACCESS_READ;
        description.MiscFlags = 0;
        com_ptr<ID3D11Texture2D> staging;
        check_hresult(d3dDevice->CreateTexture2D(&description, nullptr, staging.put()));

        D3D11_BOX region{0, 0, 0, description.Width, description.Height, 1};
        context->CopySubresourceRegion(staging.get(), 0, 0, 0, 0, source.get(), 0, &region);
        D3D11_MAPPED_SUBRESOURCE mapped{};
        check_hresult(context->Map(staging.get(), 0, D3D11_MAP_READ, 0, &mapped));
        Image image;
        image.width = static_cast<int>(description.Width);
        image.height = static_cast<int>(description.Height);
        auto rowBytes = static_cast<std::size_t>(image.width) * 4;
        image.bgra.resize(rowBytes * static_cast<std::size_t>(image.height));
        for (int row = 0; row < image.height; ++row) {
            std::memcpy(
                image.bgra.data() + static_cast<std::size_t>(row) * rowBytes,
                static_cast<std::uint8_t*>(mapped.pData) + static_cast<std::size_t>(row) * mapped.RowPitch,
                rowBytes);
        }
        context->Unmap(staging.get(), 0);
        frame.Close();
        result = std::move(image);
        if (nonBlackTenths(*result) >= 20) break;
    }
    pool.FrameArrived(token);
    session.Close();
    pool.Close();
    return result;
}

std::optional<Image> captureWithPrintWindow(HWND window) {
    RECT rect{};
    if (!GetWindowRect(window, &rect)) return std::nullopt;
    int width = rect.right - rect.left;
    int height = rect.bottom - rect.top;
    if (width <= 0 || height <= 0) return std::nullopt;

    HDC screen = GetDC(nullptr);
    HDC memory = CreateCompatibleDC(screen);
    BITMAPINFO info{};
    info.bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
    info.bmiHeader.biWidth = width;
    info.bmiHeader.biHeight = -height;
    info.bmiHeader.biPlanes = 1;
    info.bmiHeader.biBitCount = 32;
    info.bmiHeader.biCompression = BI_RGB;
    void* bits = nullptr;
    HBITMAP bitmap = CreateDIBSection(memory, &info, DIB_RGB_COLORS, &bits, nullptr, 0);
    if (!bitmap || !bits) {
        DeleteDC(memory);
        ReleaseDC(nullptr, screen);
        return std::nullopt;
    }
    HGDIOBJ old = SelectObject(memory, bitmap);
    BOOL ok = PrintWindow(window, memory, 0x00000002);

    Image image;
    image.width = width;
    image.height = height;
    auto byteCount = static_cast<std::size_t>(width) * height * 4;
    image.bgra.assign(static_cast<std::uint8_t*>(bits), static_cast<std::uint8_t*>(bits) + byteCount);
    SelectObject(memory, old);
    DeleteObject(bitmap);
    DeleteDC(memory);
    ReleaseDC(nullptr, screen);
    if (!ok) return std::nullopt;
    return image;
}

int nonBlackTenths(Image const& image) {
    std::size_t count = 0;
    for (std::size_t index = 0; index < image.bgra.size(); index += 4) {
        if (image.bgra[index] > 8 || image.bgra[index + 1] > 8 || image.bgra[index + 2] > 8) ++count;
    }
    auto pixels = static_cast<std::size_t>(image.width) * image.height;
    return pixels ? static_cast<int>(count * 1000 / pixels) : 0;
}

bool writeBMP(std::wstring const& path, Image const& image) {
    auto rowBytes = static_cast<std::uint32_t>(image.width * 4);
    auto pixelBytes = rowBytes * static_cast<std::uint32_t>(image.height);
    BITMAPFILEHEADER file{};
    file.bfType = 0x4D42;
    file.bfOffBits = sizeof(BITMAPFILEHEADER) + sizeof(BITMAPINFOHEADER);
    file.bfSize = file.bfOffBits + pixelBytes;
    BITMAPINFOHEADER info{};
    info.biSize = sizeof(info);
    info.biWidth = image.width;
    info.biHeight = image.height;
    info.biPlanes = 1;
    info.biBitCount = 32;
    info.biCompression = BI_RGB;
    info.biSizeImage = pixelBytes;

    std::ofstream output(path, std::ios::binary);
    if (!output) return false;
    output.write(reinterpret_cast<char const*>(&file), sizeof(file));
    output.write(reinterpret_cast<char const*>(&info), sizeof(info));
    for (int row = image.height - 1; row >= 0; --row) {
        output.write(
            reinterpret_cast<char const*>(image.bgra.data() + static_cast<std::size_t>(row) * rowBytes),
            rowBytes);
    }
    return output.good();
}

} // namespace

int wmain(int argc, wchar_t** argv) {
    if (argc < 3) {
        std::fputs("usage: wincap.exe <title substring> <out.bmp>\n", stderr);
        return 2;
    }
    init_apartment(apartment_type::multi_threaded);
    SetProcessDpiAwarenessContext(DPI_AWARENESS_CONTEXT_PER_MONITOR_AWARE_V2);
    needle = lowercase(argv[1]);
    auto deadline = GetTickCount64() + 5000;
    do {
        target = nullptr;
        EnumWindows(findWindow, 0);
        if (target) break;
        Sleep(250);
    } while (GetTickCount64() < deadline);
    if (!target) {
        std::fwprintf(stderr, L"no visible window matched \"%ls\"\n", argv[1]);
        return 1;
    }

    RECT rect{};
    GetWindowRect(target, &rect);
    auto exstyle = static_cast<unsigned long long>(GetWindowLongPtrW(target, GWL_EXSTYLE));
    DWORD cloaked = 0;
    DwmGetWindowAttribute(target, DWMWA_CLOAKED, &cloaked, sizeof(cloaked));
    DWORD affinity = 0;
    GetWindowDisplayAffinity(target, &affinity);
    std::printf("window rect: (%ld,%ld)-(%ld,%ld) %ldx%ld  exstyle: 0x%llx\n",
        rect.left, rect.top, rect.right, rect.bottom,
        rect.right - rect.left, rect.bottom - rect.top, exstyle);
    std::printf("window state: cloaked=%lu displayAffinity=0x%lx\n", cloaked, affinity);

    std::optional<Image> image;
    char const* method = "WindowsGraphicsCapture";
    try {
        image = captureWithWGC(captureItemForWindow(target), std::chrono::milliseconds(750));
    } catch (hresult_error const& error) {
        std::printf("WindowsGraphicsCapture failed: 0x%08lx\n", static_cast<unsigned long>(error.code().value));
    }
    if (image) {
        auto wgcTenths = nonBlackTenths(*image);
        std::printf("WindowsGraphicsCapture frame: %dx%d  non-black: %d.%d%%\n",
            image->width, image->height, wgcTenths / 10, wgcTenths % 10);
    }
    if (!image || nonBlackTenths(*image) < 20) {
        method = "PrintWindowFallback";
        image = captureWithPrintWindow(target);
    }
    if (!image) {
        std::fputs("capture failed\n", stderr);
        return 1;
    }

    auto tenths = nonBlackTenths(*image);
    auto pixels = static_cast<std::size_t>(image->width) * image->height;
    std::size_t nonBlack = pixels * static_cast<std::size_t>(tenths) / 1000;
    std::printf("method: %s\n", method);
    std::printf("size: %dx%d  non-black: %zu/%zu (%d.%d%%)\n",
        image->width, image->height, nonBlack, pixels, tenths / 10, tenths % 10);
    if (!writeBMP(argv[2], *image)) {
        std::fwprintf(stderr, L"write failed: %ls\n", argv[2]);
        return 1;
    }
    std::wprintf(L"wrote %ls\n", argv[2]);
    if (tenths < 20) {
        std::printf("capture rejected: %d.%d%% non-black is below the 2%% minimum\n", tenths / 10, tenths % 10);
        return 3;
    }
    return 0;
}
